### ISSUE TITLE
Implement Active Inference and Free Energy Minimization

### ISSUE THREAD

**Issue by DrDrewCain** [DrDrewCain/MiniMacLLM#25]
## Overview
Implement active inference framework based on the Free Energy Principle for principled uncertainty quantification and active data selection.

## Priority
⭐⭐ (Medium Priority - Research-Oriented)

## Expected Benefits
- **Principled uncertainty**: Free energy quantifies model uncertainty
- **Active data selection**: Minimize expected free energy to select informative experiences
- **Generative replay**: Generate plausible synthetic data (vs storing raw data)
- **Hierarchical priors**: Different domains have different priors p(θ|domain)
- **Connection to EWC**: EWC is a special case of free energy with Laplace approximation

### Free Energy Principle (FEP)

Biological systems minimize surprise by:
1. **Perception**: Update internal model to explain observations
2. **Action**: Sample the world to confirm predictions
3. **Learning**: Improve generative model over time

### Variational Free Energy

```python
F = D_KL[q(θ|φ) || p(θ)] - E_q[log p(D|θ)]
    ^^^^^^^^^^^^^^^^^^^     ^^^^^^^^^^^^^^^^
       Complexity            Accuracy
```

Components:
- q(θ|φ): Variational posterior (recognition model)
- p(θ): Prior over parameters
- D: Observed data

### Active Inference Update

```python
# Perception (minimize F over q)
φ* = argmin_φ F(D, φ)

# Action (minimize expected F)
a* = argmin_a E[F(D, φ*) | a]

# Learning (minimize F over θ)
θ* = argmin_θ F(D, θ)
```

### 1. Free Energy Regularization

Add VAE-style regularization to LLM:

```python
class FreeEnergyLLM(nn.Module):
    def __init__(self, base_model):
        self.base_model = base_model
        # Recognition model q(z|x)
        self.encoder = nn.Linear(d_model, latent_dim * 2)  # μ, log σ
        # Generative model p(x|z)
        self.decoder = nn.Linear(latent_dim, d_model)
    
    def forward(self, x, labels):
        # Infer latent state
        mu, log_sigma = self.encoder(hidden).chunk(2, dim=-1)
        z = mu + exp(log_sigma) * randn_like(mu)
        
        # Predict observations
        x_recon = self.decoder(z)
        
        # Free energy = Reconstruction + KL
        recon_loss = mse_loss(x_recon, hidden.detach())
        kl_loss = -0.5 * sum(1 + 2*log_sigma - mu**2 - exp(2*log_sigma))
        
        return lm_loss + beta * (recon_loss + kl_loss)
```

### 2. Active Data Selection

Select experiences that minimize expected free energy:

```python
class ActiveContinualLearner:
    def compute_expected_free_energy(self, experience):
        # Epistemic value: information gain
        entropy = -sum(p * log(p))
        
        # Pragmatic value: expected reward
        expected_reward = softmax(logits).max()
        
        # Balance exploration and exploitation
        efe = -entropy + lambda_reward * expected_reward
        return efe
    
    def select_experiences(self, candidate_pool, budget):
        # Select experiences with lowest EFE
        efes = [compute_efe(exp) for exp in candidate_pool]
        indices = argsort(efes)[:budget]
        return [candidate_pool[i] for i in indices]
```

### 3. Generative Replay

Generate synthetic experiences using the generative model:

```python
class GenerativeFreeEnergyReplay:
    def generate_replay_samples(self, num_samples, domain):
        # Sample from prior
        z_prior = randn(num_samples, latent_dim)
        
        # Generate observations
        x_gen = gen_model(z_prior, domain=domain)
        
        # Check plausibility via free energy
        z_inferred = rec_model(x_gen)
        free_energy = mse(z_inferred, z_prior)
        
        # Keep samples with low free energy
        valid = x_gen[free_energy < threshold]
        return valid
```

## Implementation Files

Create:
- `src/neurobio/active_inference.py` - FreeEnergyRegularizer, EFE computation
- `src/continual/active_selection.py` - ActiveExperienceSelector
- `src/continual/generative_replay.py` - Generative replay with FEP
- `tests/unit/test_active_inference.py` - Unit tests
- `tests/integration/test_active_integration.py` - Integration tests

## Configuration Options

Add to `ContinualLearningConfig`:
```python
use_active_inference: bool = False  # Optional
free_energy_beta: float = 0.1  # Weight for FE term
latent_dim: int = 128  # Latent space dimension
use_active_selection: bool = False  # EFE-based selection
use_generative_replay: bool = False  # Generate vs store
selection_budget: int = 100  # Experiences per update
```

## Key Research Papers

1. **Experimental Validation**: Isomura et al., 2023. "In vitro neural networks minimise variational free energy" Nature Communications, doi:10.1038/s41467-023-40141-z
   - Quantitative confirmation of FEP in real neurons
   - Synaptic connectivity encodes generative model

2. **Neural Network Implementation**: Salvatori et al., 2023. "Brain-inspired Machine Intelligence" arXiv:2306.06792
   - Helmholtz machine + active inference
   - >99% accuracy with active inference tuning

3. **Free Energy Principle Foundation**: Friston, 2010. "The free-energy principle: a unified brain theory?" Nature Reviews Neuroscience 11:127-138

4. **Active Inference**: Friston et al., 2015. "Active inference and epistemic value" Cognitive Neuroscience 6(4):187-214

5. **Canonical Neural Networks**: Buckley et al., 2017. "The free energy principle for action and perception" arXiv:1705.09156
   - Standard neural nets perform active inference
   - Backprop approximates free energy minimization

## Challenges and Considerations

1. **Computational Cost**: Iterative inference slow (10x vs feedforward)
2. **Memory Overhead**: ~2x (store hidden + errors)
3. **Hyperparameter Sensitive**: Multiple learning rates
4. **Limited Large-Scale Demos**: Mostly vision, few language models

### Phase 1: Free Energy Regularization
1. Add VAE-style encoder/decoder
2. Compute free energy during training
3. Benchmark effect on continual learning

### Phase 2: Active Data Selection
1. Implement EFE computation
2. Replace random sampling with EFE-based
3. Measure sample efficiency

### Phase 3: Generative Replay
1. Train generative model
2. Generate synthetic experiences
3. Compare to stored replay

## Testing Requirements

1. **Unit Tests**:
   - VAE encoder/decoder
   - Free energy computation
   - EFE calculation
   - Generative sampling

2. **Integration Tests**:
   - FE regularization with LLM
   - Active selection integration
   - Generative replay quality
   - End-to-end continual learning

3. **Benchmarks**:
   - Sample efficiency
   - Uncertainty calibration
   - Generative replay quality
   - Continual learning performance

## Success Criteria

- [ ] Free energy regularization implemented
- [ ] EFE-based active selection working
- [ ] Generative replay functional
- [ ] Uncertainty well-calibrated
- [ ] Sample efficiency improved
- [ ] All tests passing
- [ ] Performance within 5% of baseline

## Connection to Existing Work

- **EWC**: Special case of free energy with Laplace approximation
- **Predictive Coding**: Shares free energy minimization objective
- **Experience Replay**: Can be replaced with generative replay

## Notes

- Consider as **future enhancement** after core mechanisms proven
- Most valuable for **active learning scenarios** with limited labels
- **Research-oriented**: Requires significant architectural changes

## References

See detailed research summary in conversation context.
